platform/andes/ae350/plmt.c

```diff
@@ -0,0 +1,95 @@
+/*
+ * SPDX-License-Identifier: BSD-2-Clause
+ *
+ * Copyright (c) 2019 Andes Technology Corporation
+ *
+ * Authors:
+ *   Zong Li <zong@andestech.com>
+ *   Nylon Chen <zong@andestech.com>
+ */
+
+#include <sbi/riscv_io.h>
```

Comment:
It looks that the programming model of this PLMT timer is compatible with CLINT, with the difference of timer register offsets. Can we refactor the existing OpenSBI clint driver a little bit, to easily support PLMT?

Reply:
I agree with Bin's comment. PLMT is indeed very similar to CLINT.

We can extend clint_cold_timer_init() to make it suitable for AE350 SOC family. The clint_cold_timer_init() will required addition parameter "type". Based on this "type" parameter we can choose appropriate register offsets. The "type" can be 0 = SiFive OR 1 = AndesTech.

Reply:
Instead of requiring an additional parameter "type" for clint_cold_timer_init(), we can add two more parameters "time_val_off" and "time_cmp_off" so that the clint timer driver can be reused for Andes PLMT timer.

Reply:
Although they look similar in this stage, we are not intended to be compatible with CLINT.  We handle SW (in plicsw.c) and timer (PLMT mentioned here) interrupts separately in Andes AE350 platform, while CLINT is responsible for them both.

Reply:
Fair enough, you can keep PLMT driver in your platform support code.

If you are going to continue with PLMT and PLICSW for future platforms as well then you can also consider moving PLMT and PLICSW to lib/utils/sys (just like CLINT).

Reply:
While that's a good suggestion, I believe we will keep AE350 that way for a long time and there will be no other platforms in any near future.  Once we have our next platforms that share such design, we will re-form the sources for a better structure.

Thanks anyway! :)

Reply:
Sure, no problem.